Develop

SergeyAseev opened the pull request:
Проект готов к проверке

SergeyAseev committed:
start work with getPopularFilmsByGenreAndYear

SergeyAseev committed:
start work with getPopularFilmsByGenreAndYear

SergeyAseev committed:
application.properties was added to develop

SergeyAseev committed:
Most popular films by genre or year are returning

SergeyAseev committed:
Most popular films by genre or year are returning

SergeyAseev committed:
Merge branch 'add-most-populars' into develop

IvanKondrashkov committed:
feat: add reviews (#9)

* feat: add reviews

* refactor: like, dislike function

* refactor: add java doc ReviewDao

* refactor: findById method in ReviewDbStorage

ArtemMironiuk committed:
Add director (#8)

* feat:add_director start 12 sprint group

* feat:add director and directorcontroller

* feat:add director finish

* fix:bug fixes

Co-authored-by: Sergey Aseev <asesergej@yandex.ru>

anatolykalugin committed:
fix: исправить методы в UserDbService и FilmDbStorage, 134/134 по тестам (#10)

Co-authored-by: Sergey Aseev <asesergej@yandex.ru>

SergeyAseev committed:
Add feed (#11)

* have to switch on another branch.
OperationEnum, EventEnum and Feed.java were added

* have to switch on another branch.
OperationEnum, EventEnum and Feed.java were added

* Ready without reviewe

* 1) add-feed for review
2) add-feed for likes
3) small refactoring

* small refactoring after review

anatolykalugin committed:
Add search (функционал Поиск) (#12)

* feat: добавить поиск по режиссерам и названиям

* refactor: небольшой косметический ремонтик

* refactor: навел красоту)

* chore: добавить описание методов в интерфейсы

MarisKisKis committed:
Add common films (#13)

* common films api added

* "common films"

* "common films"

* safe commit 2

* commit after first review

* косметическая правка

* bug fix

* nor ready. problems with exceptions

* bugfix#3...

Co-authored-by: Sergey Aseev <asesergej@yandex.ru>

SergeyAseev committed:
bugfix for develop-test

SergeyAseev committed:
bugfix for develop-tests

MarisKisKis committed:
Add recommendations (#23)

* safe commit

* commit before develop merge

* commit 01.09

* доделана ветка рекомендаций

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -105,26 +182,96 @@ public void removeLike(long filmId, long userId) {
 
+    @Override
+    public List<Film> findSortFilmsByDirector(Integer directorId, String sortBy) {
+        if (findIdDirector(directorId) == null || findIdDirector(directorId).isEmpty()) {
```
Почему 2 раза делается find? Лучше бы использовать переменные

ArtemMironiuk replied:
Сделал через переменную

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -105,26 +182,96 @@ public void removeLike(long filmId, long userId) {
+                        "left join FILMS_DIRECTORS fd ON f.id = fd.FILM_ID " +
+                        "left join DIRECTORS d ON fd.DIRECTOR_ID = d.DIRECTOR_ID WHERE fd.DIRECTOR_ID = ? " +
+                        "order by extract(year from f.RELEASE_DATE) ";
+                List<Film> films = jdbcTemplate.query(sqlQuery, this::makeFilm, directorId);
```
Это общая часть, ее можно было бы вынести от этих 2 ветвлений, чтобы не было дублирования

ArtemMironiuk replied:
не понимаю как это можно вынести? это же два разных запроса

SergeyAseev replied:
Спасибо за комментарий! Да, я понял, что можно вынести в константу, добавив ей модификатор доступа. Однако считаю, что это немного затрудняет чтение кода, так как надо смотреть sql-запрос по отдельным частям (ярко осознал, когда понял, что можно сделать еще несколько таких констант в этом же классе).

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -105,26 +182,96 @@ public void removeLike(long filmId, long userId) {
+                        rs.getString("description"),
+                        LocalDate.parse(rs.getString("release_date")),
+                        rs.getInt("duration"));
+                film.setMpa(mpaRatingDao.getFilmMpa(film.getId()));
```
Эти операции довольно странные, почему эти данные не наполняются через join запросы?
Это очень много лишних накладных запросов

SergeyAseev replied:
Данный код сохранился со сдачи ТЗ11. Все говорили, что надо вносить минимальные правки в старый код. Поэтому мы только добавили одну строчку `film.setDirectors(directorDao.getFilmDirectors(film.getId()));` Если можно, то хотелось бы оставить данный код без изменений. По тестам бывают фильмы без рейтинга/жанра. Также наставник советовал выносить все метода работы с другими сущностями в свои классы, что и было сделано.

alexache replied:
Методы работы с другими сущностями - это ок, но касается CRUD операций, а не наполнения данными запросов фильмов. то есть сейчас часть запросов делают джоины, а часть нет, и получаются пересечения. Если не хочется слишком менять, то хотя бы проверять, что уже есть данные, и делать запрос к другим дао, только если данных не хватает

SergeyAseev replied:
Готово. Добавили проверку на null

alexache commented on src/main/java/ru/yandex/practicum/filmorate/service/UserDbService.java:
```diff
@@ -70,16 +85,79 @@ public List<User> retrieveAllUsers() {
+
+
+    private List<Long> getFilmsToRecommend(long id) {
+        List<User> allUsers = retrieveAllUsers();
```
Лучше бы часть этой логики перенести на SQL запросы - сразу получить список юзеров со схожими лайками, без прохода в цикле всех юзеров, а уже вторым шагом получать рекоммендации от этих пользователей второым SQL запросом

SergeyAseev replied:
Команда не придумала адекватного метода через запросы пока что. Ест идея использовать несколько Map<>... - в первую карту записывать всех пользователей и их фильмы, которым они поставили лайка. Потом вторым проходом передавать ID пользователя, который пришел в запросе и для каждого случая считать кол-во пересечений. Т.е. тем самым у нас не будет прохода по пользователям, которые НЕ ставили лайки вообще. Для данной реализации потребуется ввести handler, который будет отвечать только за эти методы.

alexache replied:
`select f.USER_ID, count(f.FILM_ID) rate from LIKES l
	join LIKES f on (l.FILM_ID = f.FILM_ID and l.USER_ID != f.USER_ID)
	join USERS u on l.USER_ID != u.USER_ID
	where l.USER_ID = ? 
	group by f.USER_ID 
	having rate > 1
	order by rate desc
	limit 20`
Вот пример одного запроса, чтобы получить всех юзеров с похожими предпочтениями.
Остается второй запрос получить рекомендации от таких юзеров, такой запрос уже намного легче

SergeyAseev replied:
Спасибо за подсказку. Один член комманды предложил свой вариант.

alexache reviewed (changes requested):
Очень объемная и качественная работа!

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -58,8 +58,24 @@ public void removeLikeToFilm(@PathVariable long id, @PathVariable long userId) {
+        return filmService.getCommonFilms(userId, friendId);
+    }
+    @GetMapping(value = "/search")
+    public List<Film> searchFilmsByDirectorOrName(@RequestParam String query,
```
Сделано

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmDbService.java:
```diff
@@ -133,8 +131,49 @@ public void removeLike(long filmId, long userId) {
+        log.info("Передан запрос на поиск фильма по названию/режиссеру");
+        try {
+            return filmStorage.searchFilmsByDirectorOrName(query, option);
+        } catch (EmptyResultDataAccessException e) {
```
Сделано

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -105,26 +182,96 @@ public void removeLike(long filmId, long userId) {
+                "left join FILMS_DIRECTORS fd ON f.id = fd.FILM_ID " +
+                "left join DIRECTORS d ON fd.DIRECTOR_ID = d.DIRECTOR_ID ";
+        if (option.contains("director") && option.contains("title") && (option.size() == 2)) {
+            sql += "WHERE UPPER(f.NAME) REGEXP ? OR UPPER(d.DIRECTOR_NAME) REGEXP ? ORDER BY f.id DESC";
```
Сделано

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -105,26 +182,96 @@ public void removeLike(long filmId, long userId) {
+                "left join GENRE g on fgl.GENRE_ID = g.ID " +
+                "left join FILMS_DIRECTORS fd ON f.id = fd.FILM_ID " +
+                "left join DIRECTORS d ON fd.DIRECTOR_ID = d.DIRECTOR_ID ";
+        if (option.contains("director") && option.contains("title") && (option.size() == 2)) {
```
Сделано

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -64,4 +67,13 @@ public List<User> retrieveFriends(@PathVariable long id) {
     public List<User> retrieveCommonFriends(@PathVariable long id, @PathVariable long otherId) {
         return userService.retrieveCommonFriends(id, otherId);
     }
+
```
Метод на 52-й строке

anatolykalugin commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -58,8 +58,24 @@ public void removeLikeToFilm(@PathVariable long id, @PathVariable long userId) {
     }
 
```
Метод на 51-й строке

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FeedDbStorage.java:
```diff
@@ -0,0 +1,36 @@
+
+@Component("FeedDbStorage")
+@RequiredArgsConstructor
+public class FeedDbStorage  implements FeedDao{
```
Сделал так, чтобы сохранить единообразие в проекте. Данный метод переименован.

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FeedDbStorage.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public List<Feed> getFeed(long userId) {
+        String sql = "SELECT * FROM FEED where user_id = ? ;";
```
Готово

ArtemMironiuk commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -19,6 +19,7 @@ public class Film {
     private int duration;
     private List<Genre> genres = new ArrayList<>();
     private Mpa mpa;
+    private List<Director> directors;
```
Исправил

ArtemMironiuk commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/DirectorDao.java:
```diff
@@ -0,0 +1,48 @@
+     * @param id идентификатор фильма
+     * @return список режиссеров
+     */
+    List<Director> getFilmDirectors(Long id);
```
Исправил

ArtemMironiuk commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/DirectorDbStorage.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    public Director updateDirector(Director director) {
+        Director directorId = findDirectorById(director.getId());
+        if (director.getId() == directorId.getId()) {
```
Сделал просто проверку на null

ArtemMironiuk commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/DirectorDbStorage.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    public List<Director> getFilmDirectors(Long id) {
+        final String sqlQuery = "select DIRECTOR_ID from FILMS_DIRECTORS where FILM_ID = ?";
+        List<Integer> integers = jdbcTemplate.queryForList(sqlQuery,Integer.class,id);
```
Изменил

ArtemMironiuk commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/DirectorDbStorage.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Override
+    public List<Director> getFilmDirectors(Long id) {
+        final String sqlQuery = "select DIRECTOR_ID from FILMS_DIRECTORS where FILM_ID = ?";
```
Сделано

SergeyAseev committed:
Add feed (#28)

* have to switch on another branch.
OperationEnum, EventEnum and Feed.java were added

* have to switch on another branch.
OperationEnum, EventEnum and Feed.java were added

* Ready without reviewe

* 1) add-feed for review
2) add-feed for likes
3) small refactoring

* small refactoring after review

* bugfix after review

ArtemMironiuk committed:
Add director(bugfix) (#27)

* feat:add_director start 12 sprint group

* feat:add director and directorcontroller

* feat:add director finish

* fix:bug fixes

* fix:bugfix(put_test_like)_by_MironyukArtem

* fix:bugfix

Co-authored-by: Sergey Aseev <asesergej@yandex.ru>

anatolykalugin committed:
Add search (fixed) (#25)

* feat: добавить поиск по режиссерам и названиям

* refactor: небольшой косметический ремонтик

* refactor: навел красоту)

* chore: добавить описание методов в интерфейсы

* fix: изменить возвр.тип на Set, изменить regexp на like, перенести try-catch в FilmDbStorage, вынести проверку критериев поискового запроса в отдельные методы

* bugfix for branch add-search

Co-authored-by: Sergey Aseev <asesergej@yandex.ru>

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,72 @@
+
+    @PutMapping
+    public Review update(@Valid @RequestBody Review review) {
+        log.info("Send post request /reviews {}", review);
```
готово

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -93,6 +120,56 @@ public List<Film> returnTopFilms(int count) {
 
+    @Override
+    public List<Film> getCommonFilms(long userId, long friendId) {
+        String sql = "select f.* from films f where f.id = " +
```
готово

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            return reviewDao.findById(id);
+        } catch (EmptyResultDataAccessException e) {
+            throw new NotFoundException(String.format("Отзыв по id=%s не найден!", id), e);
```
Тогда получается, что проект теряет единый стиль. Если посмотреть на другие классы в папке service, то там везде кидаются аналогичные исключения. Да и сервис ничего и не знает о способе хранения данных. Он в данном случае просто говорит, что данных нет.

IvanKondrashkov replied:
Добрый день, мне кажется что обрабатывать исключения в одном слое, в данном случае сервисном, это как раз плюс, а не минус.

alexache replied:
EmptyResultDataAccessException - это ошибка из Spring DAO, то есть очень конкретный способ хранения

SergeyAseev replied:
Готово

SergeyAseev commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -64,4 +66,19 @@ public interface UserService {
+     * С помощью приватного метода getFilmsToRecommend выявляем фильмы для рекомендаций
+     * @return список список фильмов
+     */
+    List <Optional<Film>> getRecommendations(long id);
```
Готово

SergeyAseev committed:
bugfix for list<optional> films and small refactoring

anatolykalugin commented:
В результате работы над групповым проектом были добавлены такой функционал, как поиск фильма по названию или режиссеру, общие фильмы, рекомендации, лента событий, режиссеры, отзывы, удаление пользователей и фильмов, а также вывод самых популярных фильмов по жанрам и годам.

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/FilmDbStorage.java:
```diff
@@ -122,9 +123,9 @@ public List<Film> returnTopFilms(int count) {
-        String sql = "select f.* from films f where f.id = " +
-                "  (select l.FILM_ID from likes l " +
-                "where l.USER_ID= ?and  l.film_id = (select ll.FILM_ID from likes ll where ll.USER_ID=? ))";
+        String sql = "select f.* from films f " +
```
Это все равно запрос со вложенными запросами, хоть и обернутый в джоины. Вложенные запросы тут не нужны
`select * from films f 
   left join MPA m on (f.MPA_ID = m.MPA_ID
   left join DIRECTORS d on (f.DIRECTOR_ID = d.DIRECTOR_ID
   join LIKES lu on (lu.FILM_ID = f.FILM_ID)
   join LIKES lf on (lf.FILM_ID = f.FILM_ID)
   where  lu.USER_ID = ? AND  lf.USER_ID = ?'

SergeyAseev replied:
Спасибо за подсказку! Обязательно поделюсь данным запросом с командой!

alexache reviewed (changes requested):
Осталось всего немного комментариев - отеетом на ваши комментарии, в основном про SQL запросы и Spring DAO exception обработки на уровне сервиса

IvanKondrashkov committed:
refactor: reviews dao and recommendations (#30)

Забирую код в develop, чтобы протестировать все вместе.

SergeyAseev committed:
1) check for null in makeFilm's method were added
2) small refactoring

alexache commented on src/main/java/ru/yandex/practicum/filmorate/storage/dao/UserDbStorage.java:
```diff
@@ -121,14 +124,17 @@ public void removeUserById(long userId) {
+        final String sqlQueryByUser = "SELECT L.FILM_ID FROM LIKES L WHERE L.USER_ID = " + userId;
+        Set<Long> idsFilmsByUser = new HashSet<>(jdbcTemplate.queryForList(sqlQueryByUser, Long.class));
+
+        final String sqlQuery = "SELECT L.FILM_ID FROM LIKES L JOIN LIKES L2 ON L.USER_ID = L2.USER_ID AND L2.USER_ID = L.USER_ID";
```
L.USER_ID = L2.USER_ID AND L2.USER_ID = L.USER_ID - какой-то самодублирующееся условие, еще хорошо бы проверять. что здесь нет фильмов, которые первоначальный юзер уже оценивал

alexache reviewed (approved):
Мне кажется с рекоммендациями еще есть проблемы, но по сумме сторипоинтов можно сказать, что вы сделали достаточно
